source/util/move_to_front.h

```diff
@@ -0,0 +1,318 @@
+    auto it = id_to_node_.find(id);
+    assert(it != id_to_node_.end());
+    RemoveNode(it->second);
+    // The iterator should still be valid, even if RemoveNode has modified
+    // id_to_node_. But just in case erase by id, not by iterator.
+    id_to_node_.erase(id);
+
+#ifndef NDEBUG
+    deprecated_ids_.insert(id);
+#endif
+  }
+
+  // Returns the number of elements in the move-to-front sequence.
+  size_t GetSize() const {
+    return SizeOf(root_);
+  }
+
+  // The methods below are for testing only.
+
+  // Inserts the value in the internal tree data structure. For testing only.
+  void TestInsert(uint32_t val) {
+    InsertNode(CreateNode(val, val));
+  }
+
+  // Removes the value from the internal tree data structure. For testing only.
+  void TestRemove(uint32_t val) {
+    const auto it = id_to_node_.find(val);
+    assert (it != id_to_node_.end());
+    RemoveNode(it->second);
+  }
+
+  // Prints the internal tree data structure to |out|. For testing only.
+  void PrintTree(std::ostream& out, bool print_timestamp = false) const {
+    if (root_)
+      PrintTreeInternal(out, root_, 1, print_timestamp);
+  }
+
+ private:
+  // Internal tree data structure node.
+  struct Node {
+    // Fields are in order from biggest bit width to smallest.
+    // SPIR-V id.
+    uint32_t id = 0;
+    // Timestamp from a logical clock which updates every time the element is
+    // accessed.
+    uint32_t timestamp = 0;
+    // The size of the node's subtee, including the node.
+    // SizeOf(LeftOf(node)) + SizeOf(RightOf(node)) + 1.
+    uint16_t size = 0;
+    // Handles to connected nodes.
+    uint16_t left = 0;
+    uint16_t right = 0;
+    uint16_t parent = 0;
+    // Distance to the farthest leaf.
+    // Leaves have height 0, real nodes at least 1.
+    uint8_t height = 0;
+  };
+
+  // Creates node and sets correct values. Nodes should be created only through
+  // this function.
+  uint16_t CreateNode(uint32_t timestamp, uint32_t id) {
+    nodes_.emplace_back(Node());
+    Node& node = nodes_.back();
+    node.timestamp = timestamp;
+    node.id = id;
+    node.size = 1;
+    node.height = 1;
+    uint16_t handle = static_cast<uint16_t>(nodes_.size() - 1);
+    id_to_node_.emplace(id, handle);
+    return handle;
+  }
+
+  // Internal tree data structure uses handles instead of pointers. Leaves and
+  // root parent reference a singleton under handle 0. Although dereferencing
+  // a null pointer is not possible, inappropriate access to handle 0 would
+  // cause an assertion. Handles are not garbage collected if id is deprecated.
+  // But handles are recycled when a node is repositioned.
+
+  // Node accessor methods. Naming is designed to be similar to natural
+  // language as these functions tend to be used in sequences, for example:
+  // ParentOf(LeftestDescendentOf(RightOf(node)))
+
+  // Returns immutable id of the node referenced by |handle|.
+  inline uint32_t IdOf(uint16_t handle) const {
```

Comment:
I have thought it's up to the compiler to decide how to interpret the hint. Does inline then even have any use except for "it's ok for the out-of-class definition of this function to appear in multiple modules"?